server/service/src/processors/transfer/requisition/test.rs

```diff
@@ -0,0 +1,350 @@
+};
+use util::{inline_edit, inline_init, uuid::uuid};
+
+use crate::{
+    processors::test_helpers::delay_for_processor,
+    requisition::{
+        request_requisition::{UpdateRequestRequisition, UpdateRequestRequstionStatus},
+        response_requisition::{UpdateResponseRequisition, UpdateResponseRequstionStatus},
+    },
+    service_provider::ServiceProvider,
+    test_helpers::{setup_all_with_data_and_service_provider, ServiceTestContext},
+};
+
+#[actix_rt::test]
+async fn requisition_transfer() {
+    let site_id = 25;
+    let request_store_name = inline_init(|r: &mut NameRow| {
+        r.id = uuid();
+        r.name = uuid();
+    });
+
+    let request_store = inline_init(|r: &mut StoreRow| {
+        r.id = uuid();
+        r.name_id = request_store_name.id.clone();
+        r.site_id = site_id;
+    });
+
+    let response_store_name = inline_init(|r: &mut NameRow| {
+        r.id = uuid();
+        r.name = uuid();
+    });
+
+    let response_store = inline_init(|r: &mut StoreRow| {
+        r.id = uuid();
+        r.name_id = response_store_name.id.clone();
+        r.site_id = site_id;
+    });
+
+    let item1 = inline_init(|r: &mut ItemRow| {
+        r.id = uuid();
+    });
+
+    let item2 = inline_init(|r: &mut ItemRow| {
+        r.id = uuid();
+    });
+
+    let site_id_settings = inline_init(|r: &mut KeyValueStoreRow| {
+        r.id = KeyValueType::SettingsSyncSiteId;
+        r.value_int = Some(site_id);
+    });
+
+    let ServiceTestContext {
+        connection,
+        service_provider,
+        processors_task,
+        ..
+    } = setup_all_with_data_and_service_provider(
+        "requisition_transfer",
+        MockDataInserts::none().stores().names().items().units(),
+        inline_init(|r: &mut MockData| {
+            r.names = vec![request_store_name.clone(), response_store_name.clone()];
+            r.stores = vec![request_store.clone(), response_store.clone()];
+            r.items = vec![item1.clone(), item2.clone()];
+            r.key_value_store_rows = vec![site_id_settings];
+        }),
+    )
+    .await;
+
+    let test = || async move {
+        let mut tester =
+            RequisitionTransferTester::new(&request_store, &response_store, &item1, &item2);
+
+        tester.insert_request_requisition(&connection).await;
+        delay_for_processor().await;
+        tester.check_response_requisition_not_created(&connection);
+        delay_for_processor().await;
+        tester.update_request_requisition_to_sent(&service_provider);
+        delay_for_processor().await;
+        tester.check_response_requisition_created(&connection);
+        delay_for_processor().await;
+        tester.check_request_requisition_was_linked(&connection);
+        delay_for_processor().await;
+        tester.update_response_requisition_to_finalised(&service_provider);
+        delay_for_processor().await;
+        tester.check_request_requisition_status_updated(&connection);
+    };
```

Comment:
The test looks good to me.
I think this is also a valid/more common ordering of events?

```
        tester.insert_request_requisition(&connection).await;
        tester.update_request_requisition_to_sent(&service_provider);
        delay_for_processor().await; // likely to be a Sync here?
        tester.check_response_requisition_created(&connection);
        tester.update_response_requisition_to_finalised(&service_provider);
        delay_for_processor().await; // likely to be a Sync here?
        tester.check_request_requisition_was_linked(&connection);
        tester.check_request_requisition_status_updated(&connection);
```
This seems to pass fine btw but wondering if it's worth having tests for valid variations of ordering here?
I don't think it's possible to get completely out of order right?

I added some print lines to logged what's called in this test and get this, regardless of my test above vs the existing test and get the same outcome.
```
process_requisition_transfers
creating requisition response
linking requisition
process_requisition_transfers
updating requisition status
```

It was kind of surprising to me that process_requisition_transfers is only triggered twice even though there are 6 `delay_for_processor().await;`s here.

I think the `delay_for_processor().await;` after `insert_request_requisition` is not necessary for 2 reasons.
1. `insert_request_requisition` is using the `extra_mock_data` which inserts data using the repository rather than service layer.
2. Even if we had used the service layer, the insert doesn't trigger `ctx.processors_trigger.trigger_requisition_transfers();`

I would be tempted to make some changes to the code or comment to clarify the expectation here.

Ideas which would make it clearer to me would be to...
A) move `insert_request_requisition` outside the `async move` block, which would make it clearer that this is part of the test setup, rather than part of the test itself.
B) Manually call `ctx.processors_trigger.trigger_requisition_transfers();` between `insert_request_requisition` and 
`check_response_requisition_not_created` to simulate this being triggered by some other activity on the server.


BTW: I *think* that the order of proccesors in `process_requisition_transfers` is potentially important for some use cases?
I couldn't break your test by re-arranging them but wonder if it's worth adding a comment somewhere to suggest this ordering is important if it is...
```
    let processors: Vec<Box<dyn RequisitionTransferProcessor>> = vec![
        Box::new(LinkRequestRequisitionProcessor),
        Box::new(CreateResponseRequisitionProcessor),
        Box::new(UpdateRequestRequstionStatusProcessor),
    ];
```
My expectation with this ordering was that`LinkRequestRequisitionProcessor` would run first which wouldn't find a  Requisition Response record (because `CreateResponseRequisitionProcessor` hasn't run yet) so the linking would only occur on the second trigger, but that doesn't seem to be the case, I'm not entirely sure why though...

Reply:
Quick summary of in person conversation.

The ordering of the processors in this vec does define the order that they run in for each record we get from the changelog.

In this case the order is unimportant because we loop over changelog records, not processor triggers.
When `CreateResponseRequisitionProcessor` runs, a new changelog record is created which will result in the `LinkRequestRequisitionProcessor` running from the same trigger. (The next loop iteration will see the new changelog record and process it)

Reply:
I've added further diagram and comment as per our discussion (to document your comment above): e92057ac8d59702e452bc89bb6285c9c1bd350b6

Reply:
That diagram is excellent. Thanks!